lib/lltz_michelson/instruction.ml

```diff
@@ -0,0 +1,499 @@
+   The Config module is used to store the current stack and instructions and handling exeception cases where shape of the stack
+   does not match instruction requirements.
+*)
+
+module M = Michelson.Ast
+module I = M.Instruction
+module T = M.Type
+
+type t = Stack.t -> Config.t
+
+module SlotStack = Stack
+open Core
+
+let noop stack = Config.ok stack []
+
+let seq (ts : t list) : t =
+  fun stack ->
+  List.fold_left
+    ts
+    ~init:(noop stack)
+    ~f:(fun ({ instructions = instrs1; stack } as config : Config.t) t ->
+      match stack with
+      | Ok stack ->
+        let { Config.stack; instructions = instrs2 } = t stack in
+        Config.{ stack; instructions = instrs1 @ instrs2 }
+      | Exceptional -> config)
+
+let rev_prefix n l =
+  let rec aux n acc l =
+    if n = 0
+    then acc, l
+    else (
+      match l with
+      | hd :: tl -> aux (n - 1) (hd :: acc) tl
+      | [] -> acc, [])
+  in
+  aux n [] l
+
+(* https://tezos.gitlab.io/michelson-reference/#instr-DIG *)
+let dig n stack =
+  (* brings element at index n to the top *)
+  let stack =
+    match rev_prefix n stack with
+    | rev_slots1, nth :: slots2 -> List.rev_append rev_slots1 (nth :: slots2)
+    | _ ->
+      raise_s [%message "Instruction.dig: invalid stack" (stack : SlotStack.t) (n : int)]
+  in
+  Config.ok stack
+  @@
+  match n with
+  | 0 -> []
+  | 1 -> [ I.swap ]
+  | n -> [ I.dig_n n ]
+
+(* https://tezos.gitlab.io/michelson-reference/#instr-DUG *)
+let dug n stack =
+  (* n is a index into the stack, tos will be at index n after instr *)
+  let stack =
+    match List.split_n stack (n + 1) with
+    | nth :: slots1, slots2 -> slots1 @ (nth :: slots2)
+    | _ ->
+      raise_s [%message "Instruction.dig: invalid stack" (stack : SlotStack.t) (n : int)]
+  in
+  Config.ok stack
+  @@
+  match n with
+  | 0 -> []
+  | 1 -> [ I.swap ]
+  | n -> [ I.dug_n n ]
+
+(* https://tezos.gitlab.io/michelson-reference/#instr-DUP, stack is 0-indexed *)
+let dup n stack =
+  (* n is a index into the stack *)
+  let stack =
+    if List.length stack >= n
+    then `Value :: stack
+    else
+      raise_s [%message "Instruction.dup: invalid stack" (stack : SlotStack.t) (n : int)]
+  in
+  Config.ok stack [ I.dup_n (n + 1) ]
+
+(* https://tezos.gitlab.io/michelson-reference/#instr-SWAP *)
+let swap stack =
+  let stack =
+    match stack with
+    | s0 :: s1 :: slots -> s1 :: s0 :: slots
+    | _ -> raise_s [%message "Instruction.swap: invalid stack" (stack : SlotStack.t)]
+  in
+  Config.ok stack [ I.swap ]
+
+(* https://tezos.gitlab.io/michelson-reference/#instr-DROP *)
+let drop n stack =
+  (* drop's n elements *)
+  if List.length stack < n
+  then
+    raise_s [%message "Instruction.drop: invalid stack" (stack : SlotStack.t) (n : int)]
+  else (
+    match n with
+    | 0 -> Config.ok stack [] (* noop *)
+    | 1 -> Config.ok (List.tl_exn stack) [ I.drop ] (* drop just one element *)
```

Comment:
Please avoid exception-raising functions. If not possible, then handle the exception. Just grep `_exn` .